include/alpaka/workdiv/WorkDivHelpers.hpp

```diff
@@ -147,177 +150,117 @@ namespace alpaka
+            switch(gridBlockExtentSubDivRestrictions)
             {
+            case GridBlockExtentSubDivRestrictions::EqualExtent:
+                blockThreadExtent = Vec::all(core::nthRootFloor(blockThreadCountMax, TIdx{TDim::value}));
+                break;
+            case GridBlockExtentSubDivRestrictions::CloseToEqualExtent:
                 // Very primitive clipping. Just halve the largest value until it fits.
                 while(blockThreadExtent.prod() > blockThreadCountMax)
-                {
-                    auto const maxElemIdx = blockThreadExtent.maxElem();
-                    blockThreadExtent[maxElemIdx] = blockThreadExtent[maxElemIdx] / static_cast<TIdx>(2u);
-                }
-            }
-            else
-            {
-                // Very primitive clipping. Just halve the smallest value until it fits.
+                    blockThreadExtent[blockThreadExtent.maxElem()] /= TIdx{2};
+                break;
+            case GridBlockExtentSubDivRestrictions::Unrestricted:
+                // Very primitive clipping. Just halve the smallest value (which is not 1) until it fits.
                 while(blockThreadExtent.prod() > blockThreadCountMax)
                 {
-                    // Compute the minimum element index but ignore ones.
-                    // Ones compare always larger to everything else.
-                    auto const minElemIdx = static_cast<TIdx>(std::distance(
-                        &blockThreadExtent[0u],
-                        std::min_element(
-                            &blockThreadExtent[0u],
-                            &blockThreadExtent[TDim::value - 1u],
-                            [](TIdx const& a, TIdx const& b)
-                            {
-                                // This first case is redundant.
-                                /*if((a == 1u) && (b == 1u))
-                                {
-                                    return false;
-                                }
-                                else */
-                                if(a == static_cast<TIdx>(1u))
-                                {
-                                    return false;
-                                }
-                                else if(b == static_cast<TIdx>(1u))
-                                {
-                                    return true;
-                                }
-                                else
-                                {
-                                    return a < b;
-                                }
-                            })));
-                    blockThreadExtent[minElemIdx] = blockThreadExtent[minElemIdx] / static_cast<TIdx>(2u);
+                    auto const it = std::min_element(
+                        blockThreadExtent.begin(),
+                        blockThreadExtent.end() - 1, // TODO(bgruber): why omit the last element?
```

Comment:
I do not think that this is correct, the documentation of `std::min_elemet` is saying the range is `Finds the smallest element in the range [first, last).`

Reply:
I also think it's incorrect, but the previous code did:
```c++
std::min_element(
                            &blockThreadExtent[0u],
                            &blockThreadExtent[TDim::value - 1u],
                            ...
```
Which takes the address of the last element as the end of the range passed to `min_element`. And this is `.end() - 1`.

Reply:
Anyway, I added a new commit to also include the last element here. I had to update some of the reference values in the unit tests as a result. Just have a look at the last commit in isolation.

Reply:
I removed that commit again because it produced wrong test results. I also changed the comment into a doxygen comment by request from @psychocoderHPC.